src/ir/effects.h

```diff
@@ -271,10 +292,18 @@ class EffectAnalyzer {
     // anything.
     assert(!((trap && other.throws()) || (throws() && other.trap)));
     // We can't reorder an implicit trap in a way that could alter what global
-    // state is modified.
-    if ((trap && other.writesGlobalState()) ||
-        (other.trap && writesGlobalState())) {
-      return true;
+    // state is modified. However, in trapsNeverHappen mode we assume traps do
+    // not occur in practice, which lets us ignore this, at least in the case if
+    // the trap will definitely happen, since we can remove traps but not move
+    // them around, so we must check if control flow can transfer. Note that we
+    // use the property of the things being compared both executing, see the
+    // comment on this function.
+    if (!trapsNeverHappen || transfersControlFlow() ||
+        other.transfersControlFlow()) {
```

Comment:
It seems that what we're assuming is not "both analyzed code fragments are executed," but rather "both analyzed code fragments are executed unless the first transfers control." It's probably worth clarifying the opening comment and the comment on `canReorder`.

Reply:
Good point, that wasn't very clear. Updated.